### PR TITLE
Add micro-benchmarks

### PR DESCRIPTION
**Pull request by jessijzhao**
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Docs change / refactoring / dependency upgrade

## Motivation and Context / Related issue

 Add opacus microbenchmarks code #368 

## How Has This Been Tested (if it applies)

`python -m pytest tests` on a CUDA capable and CPU-only machine

### Style and Typing Tests

`black .`
`flake8 --config flake.ini`
`isort -v -l 88 -o opacus --lines-after-imports 2 -m 3 --trailing-comma  .`
`autoflake  --remove-all-unused-imports --in-place -r .`
`mypy .`

## Checklist

- [x] The documentation is up-to-date with the changes I made.
- [x] I have read the **CONTRIBUTING** document and completed the CLA (see **CONTRIBUTING**).
- [x] All tests passed, and additional code has been covered with new tests.

### CODE REVIEW

File: benchmarks/run_benchmarks.py [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+        type=int,
+        help="random seed for the first run of each layer, subsequent runs increase the random seed by 1",
+    )
+    parser.add_argument("-c", "--config_file", type=str, default="config.json")
+    parser.add_argument(
+        "--cont", action="store_true", help="only run missing experiments"
+    )
+    parser.add_argument("--no_save", action="store_true")
+    parser.add_argument("-v", "--verbose", action="store_true")
+    args = parser.parse_args()
```
Comment (ffuuugor) [inline benchmarks/run_benchmarks.py:127]: it might be useful to have a command line argument for root path for storing results

**Review by ffuuugor — COMMENTED** [review]
Amazing work, thanks a lot! We've discussed the benchmark results and what we can learn from it, but this is great from the code quality perspective too. Thanks a lot for the good Readme and test coverage.

I've left few nits through the code, and I've also noticed some lint/test failures that would be good to look at before landing:

- [CircleCI linter](https://circleci.com/gh/pytorch/opacus/9694?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-checks-link)
- When I'm running pytest locally few tests are failing. I do:

```
cd benchmarks
python -m pytest ./
```

```
========================================================================================================================================================================= short test summary info ==========================================================================================================================================================================
FAILED tests/test_layers.py::test_forward_backward[layer_list0-layer_config0] - RuntimeError: 1D target tensor expected, multi-target not supported
FAILED tests/test_layers.py::test_forward_backward[layer_list1-layer_config1] - RuntimeError: only batches of spatial targets supported (3D tensors) but got targets of dimension: 4
FAILED tests/test_layers.py::test_forward_backward[layer_list2-layer_config2] - RuntimeError: 1D target tensor expected, multi-target not supported
FAILED tests/test_layers.py::test_forward_backward[layer_list3-layer_config3] - ValueError: Expected target size (64, 64), got torch.Size([64, 256, 64])
FAILED tests/test_layers.py::test_forward_backward[layer_list4-layer_config4] - RuntimeError: 1D target tensor expected, multi-target not supported
FAILED tests/test_layers.py::test_forward_backward[layer_list5-layer_config5] - RuntimeError: 1D target tensor expected, multi-target not supported
FAILED tests/test_layers.py::test_forward_backward[layer_list6-layer_config6] - ValueError: Expected target size (64, 100), got torch.Size([64, 64, 100])
FAILED tests/test_layers.py::test_forward_backward[layer_list7-layer_config7] - ValueError: Expected target size (128, 100), got torch.Size([128, 64, 100])
========================================================================================================================================================== 8 failed, 48 passed, 412 skipped, 9 warnings in 41.21s ==========================================================================================================================================================
```

**Comment by jessijzhao** [comment]
Thanks for the suggestions, @ffuuugor! I'll address them and the Lint failure in a new commit.

Regarding the test failures, could you check your PyTorch version? I suspect you may be using an older PyTorch version <= 1.9.1. In particular, versions < 1.10.0 are missing a change in `torch.nn.functional.cross_entropy`: before PyTorch 1.10.0, `target` had to be class indices [[source](https://pytorch.org/docs/1.9.1/generated/torch.nn.functional.cross_entropy.html?highlight=cross#torch.nn.functional.cross_entropy)], whereas `F.cross_entropy` in PyTorch 1.10.0 and after also accepts class probabilities in the same shape as the `input` (i.e. the model output) [[source](https://pytorch.org/docs/1.10.0/generated/torch.nn.functional.cross_entropy.html?highlight=cross#torch.nn.functional.cross_entropy)], which makes reasoning about memory easier.

Let me know if compatibility with PyTorch 1.10.0+ is acceptable, in which case I'll add a corresponding note in the Readme.

File: benchmarks/tests/test_benchmark_layer.py [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+        forward_only: whether to benchmark forward passes or forward/backward passes
+    """
+    runtime, memory_stats = run_layer_benchmark(
+        num_repeats=num_repeats,
+        forward_only=forward_only,
+        create_layer=FakeLayer,
+        runtime=duration,
+    )
+    # account for small variations
+    assert abs(runtime - ((2 - forward_only) * duration)) < 0.01
```
Comment (ashkan-software) [inline benchmarks/tests/test_benchmark_layer.py:81]: 😎

**Review by ashkan-software — APPROVED** [review]
Thanks a lot for such an amazing PR with great documentation and high standard in coding. This looks great to me and I just left some nits. Thanks @jessijzhao!

**Comment by ffuuugor** [comment]
@jessijzhao Oh, I see, that make sense, thanks for the explanation. Upgrading PyTorch did solve the local test issues

File: benchmarks/benchmark_layer.py [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+) -> Tuple[float, Dict[str, int]]:
+    """Benchmarks a single layer for runtime and CUDA memory (if applicable).
+
+    Args:
+        num_repeats: how many times to repeat the forward(/backward) pass
+        forward_only: whether to skip the backward pass
+        create_layer: function for creating the layer, takes **kwargs
+
+    Returns:
+        Runtime as a float
```
[the remark this answers is not in the document]
Comment (jessijzhao) [reply benchmarks/benchmark_layer.py:26]: Yes, a tuple containing both runtime as a float and memory statistics as a dict (as denoted in the next line). Let me make this clearer.

[the remark this answers is not in the document]
Comment (jessijzhao) [reply benchmarks/benchmark_layer.py:26]: Yes, a tuple containing both runtime as a float and memory statistics as a dict (as denoted in the next line). Let me make this clearer.

File: benchmarks/README.md [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+    - If not batch_first: `(seq_len, batch_size, input_size)`
+    - Else: `(batch_size, seq_len, input_size)`
+
+
+## Usage
+
+If saving results, ensure that `results/raw` directory exists.
+
+```
+usage: run_benchmarks.py [-h]
```
[the remark this answers is not in the document]
Comment (jessijzhao) [reply benchmarks/README.md:73]: As mentioned on line 101, `run_benchmarks.py` without any additional arguments will replicate the results in the technical report. Let me move it up for increased visibility.

**Pull request merged**